### PR TITLE
Use kernel refs

### PR DESCRIPTION
**Pull request by theengineear**
💥 💣 🎉 👏 Jupyter Extension is using kernels by ref now.

### CODE REVIEW

File: applications/desktop/src/notebook/epics/zeromq-kernels.js [hunk 1]
```diff
@@ -148,7 +148,8 @@ export const launchKernelByNameEpic = (
             actions.launchKernel({
               kernelSpec: specs[action.payload.kernelSpecName],
               cwd: action.payload.cwd,
-              ref: action.payload.ref
+              ref: action.payload.ref,
+              selectNextKernel: action.payload.selectNextKernel
```
Comment (rgbkrk) [inline applications/desktop/src/notebook/epics/zeromq-kernels.js:152]: Excellent, I figured there'd be some passthrough here.

File: applications/jupyter-extension/nteract_on_jupyter/app/index.js [hunk 2]
```diff
@@ -27,8 +27,14 @@ function createApp(jupyterConfigData: JupyterConfigData) {
   class App extends React.Component<*> {
     notificationSystem: NotificationSystem;
 
+    componentWillMount(): void {
+      // TODO: See reducers/core/index `useCore` reducer.
+      store.dispatch(actions.useCore());
```
Comment (rgbkrk) [inline applications/jupyter-extension/nteract_on_jupyter/app/index.js:32]: 💭

File: packages/core/__tests__/epics/execute-spec.js [hunk 3]
```diff
@@ -49,11 +49,21 @@ describe("createExecuteCellStream", () => {
         return this.state;
       },
       state: {
+        core: stateModule.makeStateRecord({
+          useCore: true,
```
Comment (rgbkrk) [inline packages/core/__tests__/epics/execute-spec.js:53]: I look forward to when `useCore` is gone. 😉

File: packages/core/__tests__/epics/execute-spec.js [hunk 4]
```diff
@@ -198,11 +218,21 @@ describe("executeCellEpic", () => {
...
+            kernels: stateModule.makeKernelsRecord({
+              byRef: Immutable.Map({
+                fake: stateModule.makeRemoteKernelRecord({
+                  channels: null,
+                  status: "not connected"
+                })
+              })
+            })
+          })
+        }),
```
Comment (rgbkrk) [inline packages/core/__tests__/epics/execute-spec.js:234]: Very happy you cleaned these up to use the proper records

File: packages/core/src/actionTypes.js [hunk 5]
```diff
@@ -511,7 +515,19 @@ export type LaunchKernelAction = {
...
+};
+
+export const LAUNCH_KERNEL_BY_NAME = "LAUNCH_KERNEL_BY_NAME";
+export type LaunchKernelByNameAction = {
+  type: "LAUNCH_KERNEL_BY_NAME",
+  payload: {
+    kernelSpecName: string,
+    cwd: string,
+    ref?: KernelRef,
+    selectNextKernel: boolean
```
Comment (rgbkrk) [inline packages/core/src/actionTypes.js:530]: 👍

File: packages/core/src/components/notebook-menu/index.js [hunk 6]
```diff
@@ -20,6 +21,7 @@ type Props = {
   defaultOpenKeys?: Array<string>,
   openKeys?: Array<string>,
   cellFocused: ?string,
+  currentKernelRef: ?KernelRef,
```
Comment (rgbkrk) [inline packages/core/src/components/notebook-menu/index.js:24]: 👍

File: packages/core/src/components/notebook-menu/index.js [hunk 7]
```diff
@@ -48,6 +50,7 @@ class PureNotebookMenu extends React.Component<Props> {
   static defaultProps = {
     cellFocused: null,
     saveNotebook: null,
+    currentKernelRef: null,
```
Comment (rgbkrk) [inline packages/core/src/components/notebook-menu/index.js:53]: 👍

File: packages/core/src/components/notebook-menu/index.js [hunk 8]
```diff
@@ -185,25 +189,22 @@ class PureNotebookMenu extends React.Component<Props> {
...
+          restartKernel({ ref: currentKernelRef });
         }
         break;
       case MENU_ITEM_ACTIONS.KILL_KERNEL:
         if (killKernel) {
-          // TODO: provide KernelRef
-          killKernel({});
+          killKernel({ ref: currentKernelRef });
         }
+        break;
```
Comment (rgbkrk) [inline packages/core/src/components/notebook-menu/index.js:204]: 😊

File: packages/core/src/dummy/index.js [hunk 9]
```diff
@@ -86,6 +90,8 @@ export function dummyStore(config: *) {
   const dummyNotebook = buildDummyNotebook(config);
 
   return createStore(rootReducer, {
+    core: makeStateRecord(),
+    modals: makeModalsRecord(),
```
Comment (rgbkrk) [inline packages/core/src/dummy/index.js:94]: Hooray for a fixed dummy store!

File: packages/core/src/reducers/core/communication/kernels.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
+// @flow
+import {
+  makeKernelCommunicationRecord,
+  makeKernelsCommunicationRecord
+} from "../../../state/communication/kernels";
+import * as actionTypes from "../../../actionTypes";
+import { combineReducers } from "redux-immutable";
+import * as Immutable from "immutable";
+
+// TODO: we should spec out a way to watch the killKernel lifecycle.
```
Comment (rgbkrk) [inline packages/core/src/reducers/core/communication/kernels.js:10]: Maybe we need to set the status in the kernel object when we see a kill kernel for a given ref.

File: packages/core/src/reducers/core/entities/kernels.js [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+  makeRemoteKernelRecord,
+  makeKernelsRecord
+} from "../../../state/entities/kernels";
+import * as actionTypes from "../../../actionTypes";
+import { combineReducers } from "redux-immutable";
+import * as Immutable from "immutable";
+
+// TODO: we need to clean up references to old kernels at some point. Listening
+// for KILL_KERNEL_SUCCESSFUL seems like a good candidate, but I think you can
+// also end up with a dead kernel if that fails and you hit KILL_KERNEL_FAILED.
```
Comment (rgbkrk) [inline packages/core/src/reducers/core/entities/kernels.js:13]: True true.

File: packages/core/src/state/entities/kernels.js [hunk 12]
```diff
@@ -0,0 +1,76 @@
+// @flow
+import * as Immutable from "immutable";
+import type { ChildProcess } from "child_process";
+import type { HostRef, KernelRef } from "../refs";
+import type { KernelId } from "../ids";
+
+export type BaseKernelProps = {
```
Comment (rgbkrk) [inline packages/core/src/state/entities/kernels.js:7]: It scares me a bit that these are copied over from `old` when they're exactly the same. We should make one rely on the other or just get things aligned.

**Review by rgbkrk — APPROVED** [review]
I think the only thing that scares me in this PR is that types are copied from old in `packages/core/src/state/entities/kernels.js` when they should use the same type.

**Comment by theengineear** [comment]
>I think the only thing that scares me in this PR is that types are copied from old in packages/core/src/state/entities/kernels.js when they should use the same type.

Yah, this was honestly just because I was not really trusting flow at the time and didn't want to make more waves than necessary. We should be able to sync them though in the follow up.

File: packages/core/src/reducers/core/communication/kernels.js (hunk 10 above)
Comment (theengineear) [reply packages/core/src/reducers/core/communication/kernels.js:10]: ^^ That's a lovely idea ❤️

**Comment by rgbkrk** [comment]
> Yah, this was honestly just because I was not really trusting flow at the time and didn't want to make more waves than necessary. We should be able to sync them though in the follow up.

Works for me, I'm happy to keep this all moving forward.

File: packages/core/src/state/entities/kernels.js (hunk 12 above)
Comment (rgbkrk) [reply packages/core/src/state/entities/kernels.js:7]: On further digging, these came from the previous kernel refs PR (that we closed) and they're not consistent with the ones used in old currently.

**Pull request merged**